src/r1cs/sparse.rs

```diff
@@ -0,0 +1,259 @@
+//! `num_traits::Zero`, which is not compatiable with `ff::PrimeField`.
+
+use abomonation::Abomonation;
+use abomonation_derive::Abomonation;
+use ff::PrimeField;
+use rayon::prelude::*;
+use serde::{Deserialize, Serialize};
+
+/// CSR format sparse matrix, We follow the names used by scipy.
+/// Detailed explanation here: https://stackoverflow.com/questions/52299420/scipy-csr-matrix-understand-indptr
+#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize, Abomonation)]
+#[abomonation_bounds(where <F as PrimeField>::Repr: Abomonation)]
+pub struct SparseMatrix<F: PrimeField> {
+  #[abomonate_with(Vec<F::Repr>)]
+  /// all non-zero values in the matrix
+  pub data: Vec<F>,
+  /// column indices
+  pub indices: Vec<usize>,
+  /// row information
+  pub indptr: Vec<usize>,
+  /// number of columns
+  pub cols: usize,
+}
+
+impl<F: PrimeField> SparseMatrix<F> {
+  /// construct from Vec<usize(row), usize(col), F>
+  pub fn new(matrix: &[(usize, usize, F)], rows: usize, cols: usize) -> Self {
+    let mut new_matrix = vec![vec![]; rows];
+    for (row, col, val) in matrix {
+      new_matrix[*row].push((*col, *val));
+    }
+
+    for col in new_matrix.iter_mut() {
+      col.sort_by(|(x, _), (y, _)| x.cmp(y));
+    }
+
+    let mut indptr = vec![0; rows + 1];
+    for (i, col) in new_matrix.iter().enumerate() {
+      indptr[i + 1] = indptr[i] + col.len();
+    }
+
+    let mut indices = vec![];
+    let mut data = vec![];
+    for col in new_matrix {
+      let (idx, val): (Vec<_>, Vec<_>) = col.into_iter().unzip();
+      indices.extend(idx);
+      data.extend(val);
+    }
+
+    SparseMatrix {
+      data,
+      indices,
+      indptr,
+      cols,
+    }
+  }
+
+  /// multiply by dense vector; should use rayon/gpu
+  #[tracing::instrument(skip_all, name = "SparseMatrix::multiply_vec")]
+  pub fn multiply_vec(&self, vector: &[F]) -> Vec<F> {
+    assert_eq!(self.cols, vector.len(), "invalid shape");
+
+    // naive implementation for now
+    self
+      .indptr
+      .par_windows(2)
+      .map(|ptrs| {
+        let data = &self.data[ptrs[0]..ptrs[1]];
+        let indices = &self.indices[ptrs[0]..ptrs[1]];
+        data
+          .iter()
+          .zip(indices)
+          .map(|(val, col_idx)| *val * vector[*col_idx])
+          .sum()
+      })
+      .collect()
+  }
+
+  /// multiply by dense vector; should use rayon/gpu
+  /// we don't check dimensions
+  #[tracing::instrument(skip_all, name = "SparseMatrix::multiply_vec_unchecked")]
+  pub fn multiply_vec_unchecked(&self, vector: &[F]) -> Vec<F> {
+    // naive implementation for now
+    self
+      .indptr
+      .par_windows(2)
+      .map(|ptrs| {
+        let data = &self.data[ptrs[0]..ptrs[1]];
+        let indices = &self.indices[ptrs[0]..ptrs[1]];
+        data
+          .iter()
+          .zip(indices)
+          .map(|(val, col_idx)| *val * vector[*col_idx])
+          .sum()
+      })
+      .collect()
+  }
```

Comment:
I think this is overall not bad. The initial `assert_eq` check on the matrix length should be fine to keep.

Reply:
`multiply_vec_unchecked` ended up getting removed